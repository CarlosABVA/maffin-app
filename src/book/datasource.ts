import initSqlJs from 'sql.js';
import { DataSource } from 'typeorm';

import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from './entities';

export async function initDB(database: Uint8Array): Promise<DataSource> {
  await initSqlJs({
    // Required to load the wasm binary asynchronously. Of course, you can host it wherever you want
    locateFile: file => `https://sql.js.org/dist/${file}`,
  });

  let datasource: DataSource;

  if (database.length !== 0) {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      database,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
  } else {
    datasource = new DataSource({
      type: 'sqljs',
      synchronize: true,
      logging: false,
      entities: [Account, Book, Commodity, Price, Split, Transaction],
    });
    await datasource.initialize();
    await createEmptyBook();
  }

  return datasource;
}

async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}
