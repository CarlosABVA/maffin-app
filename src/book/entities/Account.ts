import * as v from 'class-validator';
import {
  BaseEntity,
  Column,
  Entity, JoinColumn,
  ManyToOne,
  Tree,
  TreeParent,
  TreeChildren,
  OneToMany, PrimaryColumn,
} from 'typeorm';

import type Commodity from './Commodity';
import Money from '../Money';
import type Split from './Split';

/**
 * CREATE TABLE accounts (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   name            text(2048) NOT NULL,
 *   account_type    text(2048) NOT NULL,
 *   commodity_guid  CHAR(32) NOT NULL,
 *   commodity_scu   integer NOT NULL,
 *   non_std_scu     integer NOT NULL,
 *   parent_guid     CHAR(32),
 *   code            text(2048),
 *   description     text(2048),
 *   hidden          integer NOT NULL,
 *   placeholder     integer NOT NULL
 * );
 */

@Entity('accounts')
@Tree('nested-set')
export default class Account extends BaseEntity {
  static TYPES = ['ROOT', 'ASSET', 'BANK', 'CASH', 'EQUITY', 'LIABILITY', 'INCOME', 'EXPENSE', 'MUTUAL', 'STOCK'];

  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
  @v.Length(1, 32)
    guid!: string;

  @Column({
    type: 'text',
    length: 2048,
  })
  @v.MaxLength(2048)
    name!: string;

  @Column({
    type: 'text',
    enum: Account.TYPES,
    length: 2048,
    name: 'account_type',
  })
  @v.IsIn(Account.TYPES)
    type!: string;

  @Column({
    type: 'text',
    select: false,
    default: '',
  })
    path!: string;

  @TreeParent()
  @JoinColumn({ name: 'parent_guid' })
  // Seems TreeParent doesn't let us create referencing the guid of the account
  // and needs the whole object...
    parent!: Account;

  @TreeChildren()
    children!: Account[];

  @ManyToOne('Commodity', { eager: true })
  @JoinColumn({ name: 'commodity_guid' })
    fk_commodity!: Commodity | string;

  get commodity(): Commodity {
    return this.fk_commodity as Commodity;
  }

  @OneToMany('Split', (split: Split) => split.fk_account)
    splits!: Split[];

  /**
   * Note this only works if the entity has access to splits
   */
  get total(): Money {
    if (this.type === 'ROOT') {
      // @ts-ignore
      return null;
    }

    const total = this.splits.reduce(
      (acc, split) => acc.add(
        new Money(split.quantityNum / split.quantityDenom, this.commodity.mnemonic),
      ),
      new Money(0, this.commodity.mnemonic),
    );

    if (total.isNegative()) {
      return total.multiply(-1);
    }

    return total;
  }
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Account, 'name', { value: 'Account' });
