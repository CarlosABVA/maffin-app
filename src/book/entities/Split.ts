import * as v from 'class-validator';
import {
  BaseEntity,
  Column,
  Relation,
  Entity, JoinColumn, ManyToOne, PrimaryColumn,
} from 'typeorm';

import type Account from './Account';
import type Transaction from './Transaction';

/**
 * https://wiki.gnucash.org/wiki/SQL#Tables
 *
 * CREATE TABLE splits (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   tx_guid         CHAR(32) NOT NULL,
 *   account_guid    CHAR(32) NOT NULL,
 *   memo            text(2048) NOT NULL,
 *   action          text(2048) NOT NULL,
 *   reconcile_state text(1) NOT NULL,
 *   reconcile_date  timestamp NOT NULL,
 *   value_num       integer NOT NULL,
 *   value_denom     integer NOT NULL,
 *   quantity_num    integer NOT NULL,
 *   quantity_denom  integer NOT NULL,
 *   lot_guid        CHAR(32)
 * );
*/
@Entity('splits')
export default class Split extends BaseEntity {
  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
  @v.Length(1, 32)
    guid!: string;

  @ManyToOne('Transaction')
  @JoinColumn({ name: 'tx_guid' })
    fk_transaction!: Relation<Transaction> | string;

  get transaction(): Transaction {
    return this.fk_transaction as Transaction;
  }

  @ManyToOne('Account')
  @JoinColumn({ name: 'account_guid' })
    fk_account!: Relation<Account> | string;

  get account(): Account {
    return this.fk_account as Account;
  }

  // For stock operations, action is set to Buy
  @PrimaryColumn({
    type: 'varchar',
    length: 2048,
  })
    action?: string = '';

  @Column({
    type: 'integer',
    name: 'value_num',
  })
    valueNum!: number;

  @Column({
    type: 'integer',
    name: 'value_denom',
  })
    valueDenom!: number;

  @Column({
    type: 'integer',
    name: 'quantity_num',
  })
    quantityNum!: number;

  @Column({
    type: 'integer',
    name: 'quantity_denom',
  })
    quantityDenom!: number;

  /**
   * Returns the value of the transaction in the transaction's currency
   * Let's say you have the following split:
   *
   *  <split:value>100000/100</split:value>
   *  <split:quantity>1228501/10000</split:quantity>
   *
   * where the currency of the Transaction is EUR. This means you spent 1000 EUR to buy
   * 122.8501 of whatever the split's account currency is.
   *
   * When the two currencies are the same, both value and quantity are the same.
   *
   * @returns - value in transaction's currency
   */
  get value(): number {
    return this.valueNum / this.valueDenom;
  }

  /**
   * Returns the quantity spent in the currency of the split's account.
   * Let's say you have the following split:
   *
   *  <split:value>100000/100</split:value>
   *  <split:quantity>1228501/10000</split:quantity>
   *
   * where the currency of the Transaction is EUR and the symbol of the account
   * is GOOGL. This means that you spent 1000 EUR to purchase 122.8501 GOOGL
   * stocks.
   *
   * When the two currencies are the same, both value and quantity are the same.
   *
   * @returns - quantity spent in purchasing in the split account's currency
   */
  get quantity(): number {
    return this.quantityNum / this.quantityDenom;
  }
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Split, 'name', { value: 'Split' });
