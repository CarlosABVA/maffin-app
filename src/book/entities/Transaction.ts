import { DateTime } from 'luxon';
import {
  BaseEntity,
  Column, CreateDateColumn, Entity, JoinColumn, ManyToOne, OneToMany,
  PrimaryColumn,
} from 'typeorm';
import * as v from 'class-validator';

import type Commodity from './Commodity';
import type Split from './Split';
import { DateTimeTransformer } from './transformers';

/**
 * https://wiki.gnucash.org/wiki/SQL#Tables
 *
 * CREATE TABLE transactions (
 *   guid            CHAR(32) PRIMARY KEY NOT NULL,
 *   currency_guid   CHAR(32) NOT NULL,
 *   num             text(2048) NOT NULL,
 *   post_date       timestamp NOT NULL,
 *   enter_date      timestamp NOT NULL,
 *   description     text(2048)
 * );
 */

@Entity('transactions')
export default class Transaction extends BaseEntity {
  @PrimaryColumn({
    type: 'varchar',
    length: 32,
  })
    guid!: string;

  @ManyToOne('Commodity', { eager: true })
  @JoinColumn({ name: 'currency_guid' })
    fk_currency!: Commodity | string;

  get currency(): Commodity {
    return this.fk_currency as Commodity;
  }

  @Column({
    type: 'date',
    transformer: new DateTimeTransformer(),
    name: 'post_date',
  })
    date!: DateTime;

  @CreateDateColumn({
    name: 'enter_date',
  })
    enterDate?: Date;

  @OneToMany('Split', (split: Split) => split.fk_transaction)
    splits!: Split[];

  @Column({
    type: 'text',
    length: 2048,
    default: '',
  })
  @v.MaxLength(2048)
    description!: string;
}

// https://github.com/typeorm/typeorm/issues/4714
Object.defineProperty(Transaction, 'name', { value: 'Transaction' });
