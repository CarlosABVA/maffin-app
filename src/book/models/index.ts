export { default as InvestmentAccount } from './InvestmentAccount';
