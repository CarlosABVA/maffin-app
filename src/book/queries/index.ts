export { getInvestments } from './getInvestments';
export { getAccountsWithPath } from './getAccountsWithPath';
export { getMainCurrency } from './getMainCurrency';
