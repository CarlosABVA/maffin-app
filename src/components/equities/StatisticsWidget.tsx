import React from 'react';

export type StatisticsWidgetProps = {
  title: string,
  stats: string,
  description: string,
  statsTextClass?: string,
};

export default function StatisticsWidget({
  title,
  stats,
  description,
  statsTextClass = '',
}: StatisticsWidgetProps): JSX.Element {
  return (
    <div className="bg-gunmetal-700 rounded-sm mx-6 p-6">
      <p>
        {title}
      </p>
      <p className={`${statsTextClass} text-2xl font-semibold my-3`}>{stats}</p>

      <span className="text-sm">
        {description}
      </span>
    </div>
  );
}
