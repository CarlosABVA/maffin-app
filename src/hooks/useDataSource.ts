import React from 'react';
import { DataSource } from 'typeorm';

import useBookStorage from '@/hooks/useBookStorage';
import { initDB } from '@/book/datasource';

// Not sure if this is the best way to cache the storage...
// need to research
let DATASOURCE: DataSource | null = null;

export default function useDataSource(): [DataSource | null] {
  const [bookStorage] = useBookStorage();
  const [, setDataSource] = React.useState<DataSource | null>(DATASOURCE);
  const isInitializing = React.useRef(false);

  React.useEffect(() => {
    async function load() {
      if (bookStorage && DATASOURCE == null && !isInitializing.current) {
        isInitializing.current = true;
        let start;
        let end;

        start = performance.now();
        const rawBook = await bookStorage.get();
        end = performance.now();

        start = performance.now();
        // We re-check here to avoid race conditions because of
        // downloading book data takes time. Without this
        // we have some reference errors where a component may
        // use one instance another the difference instance which
        // causes synchronization issues.
        if (DATASOURCE === null) {
          DATASOURCE = await initDB(rawBook);
          end = performance.now();
          console.log(`init dataSource: ${end - start}ms`);
        }
        setDataSource(DATASOURCE);
        isInitializing.current = false;
      }
    }

    load();
  }, [bookStorage]);

  return [DATASOURCE];
}
