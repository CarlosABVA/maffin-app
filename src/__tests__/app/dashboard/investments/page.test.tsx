import React from 'react';
import { render, act } from '@testing-library/react';
import type { DataSource } from 'typeorm';

import { Commodity } from '@/book/entities';
import InvestmentsPage from '@/app/dashboard/investments/page';
import { WeightsChartProps } from '@/components/equities/WeightsChart';
import { StatisticsWidgetProps } from '@/components/equities/StatisticsWidget';
import { InvestmentsTableProps } from '@/components/equities/InvestmentsTable';
import { DividendChartProps } from '@/components/equities/DividendChart';
import Money from '@/book/Money';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/book/queries';
import * as dataSourceHooks from '@/hooks/useDataSource';

jest.mock('@/book/queries', () => ({
  __esModule: true,
  ...jest.requireActual('@/book/queries'),
}));

jest.mock('@/hooks/useDataSource', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/useDataSource'),
}));

jest.mock('@/components/equities/WeightsChart', () => {
  function WeightsChart({ investments, totalValue }: WeightsChartProps) {
    return (
      <div className="WeightsChart" data-investments="">
        <span>{JSON.stringify(investments)}</span>
        <span>{totalValue.toString()}</span>
      </div>
    );
  }

  return WeightsChart;
});

jest.mock('@/components/equities/StatisticsWidget', () => {
  function StatisticsWidget(props: StatisticsWidgetProps) {
    return (
      <div className="StatisticsWidget">
        <span>{props.description}</span>
        <span>{props.title}</span>
        <span>{props.stats}</span>
        <span>{props.description}</span>
      </div>
    );
  }

  return StatisticsWidget;
});

jest.mock('@/components/equities/InvestmentsTable', () => {
  function InvestmentsTable({ investments }: InvestmentsTableProps) {
    return (
      <div className="InvestmentsTable">
        <span>{JSON.stringify(investments)}</span>
      </div>
    );
  }

  return InvestmentsTable;
});

jest.mock('@/components/equities/DividendChart', () => {
  function DividendChart({ investments }: DividendChartProps) {
    return (
      <div className="DividendChart">
        <span>{JSON.stringify(investments)}</span>
      </div>
    );
  }

  return DividendChart;
});

describe('InvestmentsPage', () => {
  it('renders while loading data', () => {
    const { container } = render(
      <InvestmentsPage />,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders with data', async () => {
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue([{} as DataSource]);
    jest.spyOn(queries, 'getMainCurrency').mockResolvedValue({
      mnemonic: 'EUR',
      guid: 'eur',
      namespace: 'CURRENCY',
    } as Commodity);
    jest.spyOn(queries, 'getInvestments').mockResolvedValue([
      {
        account: { name: 'Investment' },
        quantity: new Money(10, 'TICKER'),
        valueInCurrency: new Money(5, 'EUR'),
        costInCurrency: new Money(4, 'EUR'),
        realizedProfitInCurrency: new Money(10, 'EUR'),
        realizedDividendsInCurrency: new Money(5, 'EUR'),
      } as InvestmentAccount,
    ]);

    let container;
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      ({ container } = render(
        <InvestmentsPage />,
      ));
    });

    expect(container).toMatchSnapshot();
  });

  it('renders with different mainCurrency', async () => {
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue([{} as DataSource]);
    jest.spyOn(queries, 'getMainCurrency').mockResolvedValue({
      mnemonic: 'USD',
      guid: 'usd',
      namespace: 'CURRENCY',
    } as Commodity);
    jest.spyOn(queries, 'getInvestments').mockResolvedValue([
      {
        account: { name: 'Investment' },
        quantity: new Money(10, 'TICKER'),
        valueInCurrency: new Money(5, 'USD'),
        costInCurrency: new Money(4, 'USD'),
        realizedProfitInCurrency: new Money(10, 'USD'),
        realizedDividendsInCurrency: new Money(5, 'USD'),
      } as InvestmentAccount,
    ]);

    let container;
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      ({ container } = render(
        <InvestmentsPage />,
      ));
    });

    expect(container).toMatchSnapshot();
  });
});
