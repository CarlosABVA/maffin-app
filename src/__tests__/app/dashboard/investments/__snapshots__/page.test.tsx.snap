// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvestmentsPage renders while loading data 1`] = `
<div>
  <span
    class="text-xl font-medium pb-4"
  >
    Your Investments
  </span>
  <div
    class="grid grid-cols-12"
  >
    <div
      class="col-span-4"
    >
      <div
        class="WeightsChart"
        data-investments=""
      >
        <span>
          []
        </span>
        <span>
          0.00 EUR
        </span>
      </div>
    </div>
    <div
      class="col-span-8"
    >
      <div
        class="grid grid-cols-12"
      >
        <div
          class="col-span-4"
        >
          <div
            class="StatisticsWidget"
          >
            <span>
              0 total invested
            </span>
            <span>
              Value/Cost
            </span>
            <span>
              0.00 €
            </span>
            <span>
              0 total invested
            </span>
          </div>
        </div>
        <div
          class="col-span-4"
        >
          <div
            class="StatisticsWidget"
          >
            <span>
              0 (NaN%) with dividends
            </span>
            <span>
              Unrealized Profit
            </span>
            <span>
              0.00 € (NaN%)
            </span>
            <span>
              0 (NaN%) with dividends
            </span>
          </div>
        </div>
        <div
          class="col-span-4"
        >
          <div
            class="StatisticsWidget"
          >
            <span>
              0.00 € from dividends
            </span>
            <span>
              Realized
            </span>
            <span>
              0.00 €
            </span>
            <span>
              0.00 € from dividends
            </span>
          </div>
        </div>
      </div>
      <div
        class="DividendChart"
      >
        <span>
          []
        </span>
      </div>
    </div>
  </div>
  <div
    class="py-4"
  >
    <div
      class="InvestmentsTable"
    >
      <span>
        []
      </span>
    </div>
  </div>
</div>
`;

exports[`InvestmentsPage renders with data 1`] = `
<div>
  <span
    class="text-xl font-medium pb-4"
  >
    Your Investments
  </span>
  <div
    class="grid grid-cols-12"
  >
    <div
      class="col-span-4"
    >
      <div
        class="WeightsChart"
        data-investments=""
      >
        <span>
          [{"account":{"name":"Investment"},"quantity":{"_raw":{"amount":10,"currency":{"code":"TICKER","base":10,"exponent":2},"scale":0}},"valueInCurrency":{"_raw":{"amount":5,"currency":{"code":"EUR","base":10,"exponent":2},"scale":0}},"costInCurrency":{"_raw":{"amount":4,"currency":{"code":"EUR","base":10,"exponent":2},"scale":0}},"realizedProfitInCurrency":{"_raw":{"amount":10,"currency":{"code":"EUR","base":10,"exponent":2},"scale":0}},"realizedDividendsInCurrency":{"_raw":{"amount":5,"currency":{"code":"EUR","base":10,"exponent":2},"scale":0}}}]
        </span>
        <span>
          5.00 EUR
        </span>
      </div>
    </div>
    <div
      class="col-span-8"
    >
      <div
        class="grid grid-cols-12"
      >
        <div
          class="col-span-4"
        >
          <div
            class="StatisticsWidget"
          >
            <span>
              4 total invested
            </span>
            <span>
              Value/Cost
            </span>
            <span>
              5.00 €
            </span>
            <span>
              4 total invested
            </span>
          </div>
        </div>
        <div
          class="col-span-4"
        >
          <div
            class="StatisticsWidget"
          >
            <span>
              6 (150%) with dividends
            </span>
            <span>
              Unrealized Profit
            </span>
            <span>
              1.00 € (25%)
            </span>
            <span>
              6 (150%) with dividends
            </span>
          </div>
        </div>
        <div
          class="col-span-4"
        >
          <div
            class="StatisticsWidget"
          >
            <span>
              5.00 € from dividends
            </span>
            <span>
              Realized
            </span>
            <span>
              15.00 €
            </span>
            <span>
              5.00 € from dividends
            </span>
          </div>
        </div>
      </div>
      <div
        class="DividendChart"
      >
        <span>
          [{"account":{"name":"Investment"},"quantity":{"_raw":{"amount":10,"currency":{"code":"TICKER","base":10,"exponent":2},"scale":0}},"valueInCurrency":{"_raw":{"amount":5,"currency":{"code":"EUR","base":10,"exponent":2},"scale":0}},"costInCurrency":{"_raw":{"amount":4,"currency":{"code":"EUR","base":10,"exponent":2},"scale":0}},"realizedProfitInCurrency":{"_raw":{"amount":10,"currency":{"code":"EUR","base":10,"exponent":2},"scale":0}},"realizedDividendsInCurrency":{"_raw":{"amount":5,"currency":{"code":"EUR","base":10,"exponent":2},"scale":0}}}]
        </span>
      </div>
    </div>
  </div>
  <div
    class="py-4"
  >
    <div
      class="InvestmentsTable"
    >
      <span>
        [{"account":{"name":"Investment"},"quantity":{"_raw":{"amount":10,"currency":{"code":"TICKER","base":10,"exponent":2},"scale":0}},"valueInCurrency":{"_raw":{"amount":5,"currency":{"code":"EUR","base":10,"exponent":2},"scale":0}},"costInCurrency":{"_raw":{"amount":4,"currency":{"code":"EUR","base":10,"exponent":2},"scale":0}},"realizedProfitInCurrency":{"_raw":{"amount":10,"currency":{"code":"EUR","base":10,"exponent":2},"scale":0}},"realizedDividendsInCurrency":{"_raw":{"amount":5,"currency":{"code":"EUR","base":10,"exponent":2},"scale":0}}}]
      </span>
    </div>
  </div>
</div>
`;

exports[`InvestmentsPage renders with different mainCurrency 1`] = `
<div>
  <span
    class="text-xl font-medium pb-4"
  >
    Your Investments
  </span>
  <div
    class="grid grid-cols-12"
  >
    <div
      class="col-span-4"
    >
      <div
        class="WeightsChart"
        data-investments=""
      >
        <span>
          [{"account":{"name":"Investment"},"quantity":{"_raw":{"amount":10,"currency":{"code":"TICKER","base":10,"exponent":2},"scale":0}},"valueInCurrency":{"_raw":{"amount":5,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}},"costInCurrency":{"_raw":{"amount":4,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}},"realizedProfitInCurrency":{"_raw":{"amount":10,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}},"realizedDividendsInCurrency":{"_raw":{"amount":5,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}}}]
        </span>
        <span>
          5.00 USD
        </span>
      </div>
    </div>
    <div
      class="col-span-8"
    >
      <div
        class="grid grid-cols-12"
      >
        <div
          class="col-span-4"
        >
          <div
            class="StatisticsWidget"
          >
            <span>
              4 total invested
            </span>
            <span>
              Value/Cost
            </span>
            <span>
              5.00 $
            </span>
            <span>
              4 total invested
            </span>
          </div>
        </div>
        <div
          class="col-span-4"
        >
          <div
            class="StatisticsWidget"
          >
            <span>
              6 (150%) with dividends
            </span>
            <span>
              Unrealized Profit
            </span>
            <span>
              1.00 $ (25%)
            </span>
            <span>
              6 (150%) with dividends
            </span>
          </div>
        </div>
        <div
          class="col-span-4"
        >
          <div
            class="StatisticsWidget"
          >
            <span>
              5.00 $ from dividends
            </span>
            <span>
              Realized
            </span>
            <span>
              15.00 $
            </span>
            <span>
              5.00 $ from dividends
            </span>
          </div>
        </div>
      </div>
      <div
        class="DividendChart"
      >
        <span>
          [{"account":{"name":"Investment"},"quantity":{"_raw":{"amount":10,"currency":{"code":"TICKER","base":10,"exponent":2},"scale":0}},"valueInCurrency":{"_raw":{"amount":5,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}},"costInCurrency":{"_raw":{"amount":4,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}},"realizedProfitInCurrency":{"_raw":{"amount":10,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}},"realizedDividendsInCurrency":{"_raw":{"amount":5,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}}}]
        </span>
      </div>
    </div>
  </div>
  <div
    class="py-4"
  >
    <div
      class="InvestmentsTable"
    >
      <span>
        [{"account":{"name":"Investment"},"quantity":{"_raw":{"amount":10,"currency":{"code":"TICKER","base":10,"exponent":2},"scale":0}},"valueInCurrency":{"_raw":{"amount":5,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}},"costInCurrency":{"_raw":{"amount":4,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}},"realizedProfitInCurrency":{"_raw":{"amount":10,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}},"realizedDividendsInCurrency":{"_raw":{"amount":5,"currency":{"code":"USD","base":10,"exponent":2},"scale":0}}}]
      </span>
    </div>
  </div>
</div>
`;
